cirq/protocols/unitary.py

```diff
@@ -0,0 +1,88 @@
+# Copyright 2018 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Any, Optional, TypeVar, Union
+
+import numpy as np
+from typing_extensions import Protocol
+
+from cirq import abc, extension
+from cirq.ops.gate_features import KnownMatrix
+
+# This is a special indicator value used by the unitary method to determine
+# whether or not the caller provided a 'default' argument. It must be of type
+# np.ndarray to ensure the method has the correct type signature in that case.
+# It is checked for using `is`, so it won't have a false positive if the user
+# provides a different np.array([]) value.
+RaiseTypeErrorIfNotProvided = np.array([])
+
+TDefault = TypeVar('TDefault')
+
+
+class SupportsUnitary(Protocol):
+    """An object that may be describable by a unitary matrix."""
+
+    @abc.abstractmethod
```

Comment:
The [protocol examples](https://mypy.readthedocs.io/en/latest/protocols.html#defining-subprotocols-and-subclassing-protocols) in the mypy docs do not mark methods as `abstractmethod`; is there a reason you've done so here? This might mean that if you _explicitly_ inherit from `SupportsUnitary` instead of just relying on ducktyping, then when the method is not present you'll get an error at instance creation, rather than later when calling `cirq.unitary`. If that in fact works and we want to support explicit inheritance, we should document that and add some tests.

Reply:
This made more sense when there were multiple methods and implementing one would make the others work. Now it makes so sense. Removed the abstractmethod.